### PR TITLE
Lazily import statsmodels and plugins

### LINKED ISSUE #1047: The RAVEN start time is too slow
--------
Issue Description
--------
##### What did you expect to see happen?
The RAVEN startup time is too slow. This is mostly due to the numerous imports we do of heavy libraries (E.g. Tensorflow, etc.), even if for most of the cases those imports are not necessaries.

##### Do you have a suggested fix for the development team?
Addition of a lazy import functionality in order to import libraries at will (when they are needed) and not at the start time.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
This lazily imports statsmodels and plugins to speed up RAVEN when they are not used.

Note that statsmodels.api needs to be imported to get other modules
in statsmodels to be usable.

--------
Pull Request Description
--------
##### What issue does this change request address? 

Closes #1047

##### What are the significant changes in functionality due to this change request?
Imports statsmodels, sklearn and plugins lazily.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by joshua-cogliati-inl** [comment]
I think the next thing to lazily import in the pull is plugins.

**Comment by alfoa** [comment]
Josh,
this is the issue that should be closed:

#1047

Please, when you are done, refer (close) to it in the description of he PR

**Comment by alfoa** [comment]
Josh I added the issue... #1047 ...should it be closed or just referenced?

File: framework/Metrics/SklMetric.py [hunk 1]
```diff
@@ -110,8 +111,9 @@ def _localReadMoreXML(self,xmlNode):
     if self.metricType[0] not in self.__class__.availMetrics.keys() or self.metricType[1] not in self.__class__.availMetrics[self.metricType[0]].keys():
       self.raiseAnError(IOError, "Metric '", self.name, "' with metricType '", self.metricType[0], "|", self.metricType[1], "' is not valid!")
 
-    if self.metricType[0] == 'paired_distance' and int(sklearn.__version__.split(".")[1]) < 18:
-      self.raiseAnError(IOError, "paired_distance is not supported in your SciKit-Learn version, if you want to use this metric, please make sure your SciKit-Learn version >= 18!")
+    # Are we supporting sklearn < 0.18 anymore?
```
Comment (alfoa) [inline framework/Metrics/SklMetric.py]: no. can you remove the if statement? and the commented part?

File: framework/SupervisedLearning/SciKitLearn.py [hunk 2]
```diff
@@ -170,6 +76,103 @@ def __init__(self,messageHandler,**kwargs):
       @ Out, None
     """
     supervisedLearning.__init__(self,messageHandler,**kwargs)
+    import sklearn # = import_module_lazy("sklearn")
```
Comment (alfoa) [inline framework/SupervisedLearning/SciKitLearn.py]: can you remove the commented code?

File: framework/utils/cached_ndarray.py [hunk 3]
```diff
@@ -24,8 +24,8 @@
 import threading
 from numpy import ndarray
 import numpy as np
-import xarray as xr
-import pandas as pd
+#import xarray as xr
+#import pandas as pd
```
Comment (alfoa) [inline framework/utils/cached_ndarray.py]: can you remove it?

File: framework/Metrics/SklMetric.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply framework/Metrics/SklMetric.py]: Removed.

File: framework/SupervisedLearning/SciKitLearn.py (hunk 2 above)
Comment (joshua-cogliati-inl) [reply framework/SupervisedLearning/SciKitLearn.py]: Done.

File: framework/utils/cached_ndarray.py (hunk 3 above)
Comment (joshua-cogliati-inl) [reply framework/utils/cached_ndarray.py]: Done.

**Comment by joshua-cogliati-inl** [comment]
> Josh I added the issue... #1047 ...should it be closed or just referenced?

It can probably be closed.  I don't think there are many more large savings to gained from future work.

**Comment by alfoa** [comment]
Checklist passed...

PR is approved.

**Pull request merged**